feat(Run): Allow teachers to archive runs

geoffreykwan opened the pull request:
## Changes

- Teachers can now archive runs
- Removed filter by period select drop down in Class Schedule view
- There is now an active runs view and an archived runs view

## Test

- Test with https://github.com/WISE-Community/WISE-API/pull/225

- Archive/unarchive a run using the run item menu list (the 3 dots menu on a run)
- Archive/unarchive a run using the run item checkboxes and then the "Archive Selected"/"Unarchive Selected" button
- Archive/unarchive a run that was shared with you
- Make sure the select all runs checkbox works properly
- Make sure the select runs drop down works properly for (all, none, running, completed)
- Make sure switching from the active to archived view works properly

Closes #1012

geoffreykwan committed:
feat(Run): Allow teachers to archive runs

#1012

geoffreykwan committed:
chore(Run): Change order of run display to completed, running, scheduled

#1012

geoffreykwan committed:
feat(Run): Change archiving to use tags instead of isDeleted field

#1012

geoffreykwan committed:
feat(Run): Move tags from run to project and use project id for archiving

#1012

geoffreykwan committed:
test(Archive): Fix archive run tests

#1012

geoffreykwan committed:
Merge branch 'develop' into issue-1012-allow-teacher-to-archive-runs

geoffreykwan committed:
feat(Archive): Allow archiving runs that are shared with you

#1012

geoffreykwan committed:
feat(Archive): Fix select runs drop down

#1012

geoffreykwan committed:
test(Archive): Fix test

#1012

geoffreykwan committed:
feat(Archive): Create component for select runs checkbox and drop down

#1012

geoffreykwan committed:
test(Archive): Add test harnesses to teacher run list tests

#1012

geoffreykwan committed:
test(Archive Run): Clean up test harnesses

#1012

geoffreykwan committed:
feat(Archive Run): Change http requests to use new methods and paths

#1012

geoffreykwan committed:
chore(Archive Run): Create ArchiveProjectService and move functions in

#1012

geoffreykwan committed:
feat(Archive Run): Archive requests now look at project status in response

#1012

geoffreykwan committed:
chore(Archive): Change boolean variable names to not start with is

#1012

hirokiterashima commented on src/app/teacher/teacher-run-list/teacher-run-list.component.html:
```diff
@@ -58,23 +58,59 @@
     </p>
   </div>
-
+  <div
+    *ngIf="filteredRuns.length > 0"
+    class="select-all-controls"
+    fxLayout="row"
+    fxLayoutAlign="start center"
+    fxLayoutGap="20px"
+  >
+    <select-runs-controls
+      [numSelectedRuns]="numSelectedRuns"
+      [numTotalRuns]="filteredRuns.length"
+      (selectRunsOptionChosenEvent)="selectRunsOptionChosen($event)"
+    ></select-runs-controls>
+    <div *ngIf="numSelectedRuns > 0" class="num-selected-runs-display" i18n>
+      {{ numSelectedRuns }} Run(s) Selected
+    </div>
+    <button
+      mat-raised-button
+      color="accent"
+      *ngIf="!showArchived && numSelectedRuns > 0"
+      (click)="archiveSelectedRuns()"
+      i18n
+    >
+      Archive Selected
+    </button>
+    <button
+      mat-raised-button
+      color="accent"
+      *ngIf="showArchived && numSelectedRuns > 0"
+      (click)="unarchiveSelectedRuns()"
+      i18n
+    >
+      Unarchive Selected
+    </button>
+  </div>
```
TeacherRunListComponent is getting big. Can we make this block into a new component?

hirokiterashima commented on src/app/teacher/run-menu/run-menu.component.ts:
```diff
@@ -86,4 +90,32 @@ export class RunMenuComponent implements OnInit {
     }
   }
+
+  archive(): void {
+    const run = this.run;
+    this.archiveProjectService.archiveProject(run.project).subscribe({
+      next: (response: ArchiveProjectResponse) => {
+        run.archived = response.archived;
+        this.runArchiveStatusChangedEvent.emit(run);
+        this.snackBar.open($localize`Successfully Archived Run`);
+      },
+      error: () => {
+        this.snackBar.open($localize`Error Archiving Run`);
+      }
+    });
+  }
+
+  unarchive(): void {
+    const run = this.run;
+    this.archiveProjectService.unarchiveProject(run.project).subscribe({
+      next: (response: ArchiveProjectResponse) => {
+        run.archived = response.archived;
+        this.runArchiveStatusChangedEvent.emit(run);
+        this.snackBar.open($localize`Successfully Unarchived Run`);
+      },
+      error: () => {
+        this.snackBar.open($localize`Error Unarchiving Run`);
+      }
+    });
+  }
```
Duplication. Can you try refactoring it?

hirokiterashima commented on src/app/services/archive-project.service.ts:
```diff
@@ -0,0 +1,32 @@
+    return this.http.delete<any>(`/api/projects/archived`, {
+      params: params
+    });
+  }
```
Instead of ```Observable<any>```, is it ```Observable<ArchiveProjectResponse[]>```?

hirokiterashima reviewed (commented):
I added some comments. Also, should we use "Archive/Unarchive Unit" instead of "Archive/Unarchive Run"? We want teachers to think in "unit", not "run"?

geoffreykwan committed:
chore(Teacher Run List): Add private/protected modifiers to variables

#1012

geoffreykwan committed:
chore(Teacher Run List Item): Add private/protected modifiers to variables

#1012

geoffreykwan committed:
chore(Select Runs Controls): Add private/protected modifiers to variables

#1012

geoffreykwan committed:
chore(Run Menu): Add private/protected modifiers and create test harness

#1012

geoffreykwan committed:
chore(Run Menu): Remove duplication in archive()/unarchive()

#1012

geoffreykwan committed:
chore(Archive Run): Change snackbar wording to use Unit instead of Run

#1012

breity committed:
Update run menu text and snackbar messages
- Change 'unarchive' to 'restore'
- Update archive and restore font icons

breity committed:
Update teacher run list item styles
- Reduce space around checkbox and run menu trigger
- Make run title warn color for archived runs
- Add tooltip and aria-label to checkbox

breity committed:
Replace active/archived slide toggle with select

breity committed:
Update run selection tools
- Replace archive/restore action button with icon button
- Update some text for brevity/clarity
- Add aria-labels and tooltips
- Clean up layout styles

breity committed:
Merge remote-tracking branch 'refs/remotes/origin/issue-1012-allow-teacher-to-archive-runs' into issue-1012-allow-teacher-to-archive-runs

breity committed:
Fix tests, update some text and function names

breity reviewed (changes requested):
Looking great. I updated styles and layout, as well as some text, accessibility elements, etc.
@geoffreykwan There are two failing tests that I'm having trouble fixing. Maybe we could take a look.

I have 1 suggestion: We should add a confirmation dialog when user chooses to archive or restore multiple runs at once. @hirokiterashima and @geoffreykwan, what do you think?

These are the issues I found that we discussed:
1. Selecting all the runs using the top checkbox doesn't select shared runs.
2. All the user's runs are shown now when loading the run list. We used to only show the latest 10 and let users view the rest with the 'Show All' button.

hirokiterashima commented:
New style changes look good. 

I wonder if asking users to confirm each time is a bit disruptive, especially since it's easy to undo their changes in the archive/unarchive views if they make a mistake by accident. How about adding an 'undo' button in the toast message, like in Gmail?

geoffreykwan committed:
fix(Archive): Do not show all and select all includes shared runs

#1012

geoffreykwan committed:
chore(Archive Run): Move archive button into select run controls

#1012

geoffreykwan committed:
feat(Archive Run): Allow undoing an archive action

#1012

hirokiterashima commented:
The new undo feature works great.

Should we hide or disable the SelectRunsControls if there are no runs to select? It seems a bit strange to have it there.
<img width="1231" alt="Screenshot 2023-08-28 at 11 20 42 AM" src="https://github.com/WISE-Community/WISE-Client/assets/119416/d1ab88f9-c66b-4e57-82c0-c4645e3431d3">

breity commented:
> Should we hide or disable the SelectRunsControls if there are no runs to select? It seems a bit strange to have it there. 

Ah, yep. Good catch. We should hide them.
EDIT: Actually, none of that should be visible when there aren't any active runs. User should see the notice with text "Hey there! Looks like you haven't run any WISE units in your classes yet..." It's this code in TeacherRunListComponent:
```
<div class="notice" *ngIf="!runs.length && loaded">
  <p i18n>Hey there! Looks like you haven't run any WISE units in your classes yet.</p>
  <p i18n>Select "Unit Library" to find titles to use with your students.</p>
</div>
<div class="content-block" *ngIf="runs.length">
...
```

We should update the *ngIf to look for any active runs, rather than any runs for the user. We should also add a different notice when viewing the archive and the user has no archived runs (i.e. "You don't have any archived classroom units.").

geoffreykwan committed:
chore(Archive Run): Change function name from restore to unarchive

#1012

geoffreykwan committed:
chore(Archive Run): Clean up code

#1012

geoffreykwan committed:
feat(Archive Run): Show message when there are no active or archived runs

#1012

hirokiterashima commented on src/app/teacher/run-menu/run-menu.component.ts:
```diff
@@ -8,23 +8,28 @@ import { ConfigService } from '../../services/config.service';
-
-  editLink: string = '';
-  reportProblemLink: string = '';
+  @Output() runArchiveStatusChangedEvent: EventEmitter<any> = new EventEmitter<any>();
```
Looks like we don't do anything with the emitted value, so make it void?
```suggestion
  @Output() runArchiveStatusChangedEvent: EventEmitter<void> = new EventEmitter<void>();
```

hirokiterashima commented on src/app/teacher/run-menu/run-menu.component.ts:
```diff
@@ -86,4 +83,57 @@ export class RunMenuComponent implements OnInit {
+
+  private updateArchivedStatus(run: TeacherRun, archived: boolean): void {
+    run.archived = archived;
+    this.runArchiveStatusChangedEvent.emit(run);
```
Looks like we don't do anything with the emitted value, so make it void?
```suggestion
    this.runArchiveStatusChangedEvent.emit();
```

hirokiterashima commented on src/app/teacher/teacher-run-list-item/teacher-run-list-item.component.ts:
```diff
@@ -15,13 +15,14 @@ import { ShareRunCodeDialogComponent } from '../share-run-code-dialog/share-run-
-  animateDuration: string = '0s';
-  animateDelay: string = '0s';
+  @Output() runArchiveStatusChangedEvent: EventEmitter<any> = new EventEmitter<any>();
+  @Output() runSelectedStatusChangedEvent: EventEmitter<any> = new EventEmitter<any>();
```
Looks like we don't do anything with the emitted value, so make it void?
```suggestion
  @Output() runArchiveStatusChangedEvent: EventEmitter<void> = new EventEmitter<void>();
  @Output() runSelectedStatusChangedEvent: EventEmitter<void> = new EventEmitter<void>();
```

hirokiterashima reviewed (changes requested):
It's getting there. As discussed this morning, here are some more changes to make:

1. In both active and archive views, add “scheduled” select-option, and show the option order as follows: All -> None -> Completed -> Running -> Scheduled

2. Change message when there are no active runs to “Hey there! Looks like you don't have any active classroom units.\n\nBrowse the "Unit Library" to find titles to use with your students.” Also remove the count message.

3. Change message when there are no archived runs to “You don’t have any archived classroom units.” Also remove the count message.

4. See minor code comments inline

geoffreykwan committed:
feat(Archive Run): Hide run count when there are no runs

Slightly change message when there are no runs.

#1012

geoffreykwan committed:
feat(Archive Run): Add scheduled option to select runs drop down

#1012

geoffreykwan committed:
chore(Archive Run): Clean up code

#1012

geoffreykwan committed:
chore(Archive Run): Clean up tests

#1012

breity committed:
Remove warn color text in archived view

geoffreykwan committed:
Merge branch 'develop' into issue-1012-allow-teacher-to-archive-runs

geoffreykwan committed:
test(Archive Run): Fix teacher-run-list-item test

#1012

hirokiterashima commented on src/app/teacher/teacher-run-list/teacher-run-list.component.spec.ts:
```diff
@@ -1,130 +1,435 @@
+      await clickSelectRunsMenuButtonAndExpectSelected('Scheduled', [true, false, false]);
+    });
+  });
+}
```
This block introduces duplication. How about writing it this way? Also, let's read and discuss [Jasmine style guide](https://github.com/CareMessagePlatform/jasmine-styleguide) to standardize how we structure/write unit tests.
```suggestion
function selectRunsOptionChosen(): void {
  describe('selectRunsOptionChosen()', () => {
    beforeEach(() => {
      setRun2Completed();
    });
    selectAllOptionChosen();
    selectNoneOptionChosen();
    selectCompletedOptionChosen();
    selectRunningOptionChosen();
    selectScheduledOptionChosen();
  });
}

function selectAllOptionChosen(): void {
  it('should select all runs', async () => {
    await clickSelectRunsMenuButtonAndExpectSelected('All', [true, true, true]);
  });
}

function selectNoneOptionChosen(): void {
  it('should select no runs', async () => {
    await clickSelectRunsMenuButtonAndExpectSelected('None', [false, false, false]);
  });
}

function selectCompletedOptionChosen(): void {
  it('should select completed runs', async () => {
    await clickSelectRunsMenuButtonAndExpectSelected('Completed', [false, true, false]);
  });
}

function selectRunningOptionChosen(): void {
  it('should select running runs', async () => {
    await clickSelectRunsMenuButtonAndExpectSelected('Running', [true, false, true]);
  });
}

function selectScheduledOptionChosen(): void {
  it('it should select scheduled runs', async () => {
    setRun3Scheduled();
    await clickSelectRunsMenuButtonAndExpectSelected('Scheduled', [true, false, false]);
  });
}
```

hirokiterashima commented on src/app/teacher/teacher-run-list/teacher-run-list.component.ts:
```diff
@@ -212,8 +200,51 @@ export class TeacherRunListComponent implements OnInit {
+    }
+  }
+
+  protected selectRunsOptionChosen(value: string): void {
+    this.filteredRuns.forEach((run: TeacherRun) => {
+      switch (value) {
+        case 'all':
+          run.selected = true;
+          break;
+        case 'none':
+          run.selected = false;
+          break;
+        case 'completed':
+          run.selected = run.isCompleted(this.configService.getCurrentServerTime());
+          break;
+        case 'running':
+          run.selected = run.isActive(this.configService.getCurrentServerTime());
+          break;
+        case 'scheduled':
+          run.selected = run.isScheduled(this.configService.getCurrentServerTime());
+          break;
+      }
+    });
+    this.runSelectedStatusChanged();
+  }
```
Put the logic in TeacherRun?
```suggestion
  protected selectRunsOptionChosen(option: SelectRunsOption): void {
    const now = this.configService.getCurrentServerTime();
    this.filteredRuns.forEach((run: TeacherRun) => run.updateSelected(option, now));
    this.runSelectedStatusChanged();
  }
```

hirokiterashima commented on src/app/teacher/teacher-run.ts:
```diff
@@ -1,8 +1,10 @@
 import { Run } from '../domain/run';
 
 export class TeacherRun extends Run {
```
Move code for setting teacherRun.selected into this class?
```
  updateSelected(selectRunsOption: SelectRunsOption, currentTime: number): void {
    switch (selectRunsOption) {
      case 'ALL':
        this.selected = true;
        break;
      case 'NONE':
        this.selected = false;
        break;
      case 'COMPLETED':
        this.selected = this.isCompleted(currentTime);
        break;
      case 'RUNNING':
        this.selected = this.isActive(currentTime);
        break;
      case 'SCHEDULED':
        this.selected = this.isScheduled(currentTime);
        break;
    }
  }
```

hirokiterashima commented on src/app/teacher/select-runs-controls/select-runs-controls.component.ts:
```diff
@@ -0,0 +1,126 @@
+  @Input() runs: TeacherRun[] = [];
+  protected selectedAllRuns: boolean = false;
+  protected selectedSomeRuns: boolean = false;
+  @Output() selectRunsOptionChosenEvent = new EventEmitter<string>();
```
Emit SelectRunsOptions enum class?
```suggestion
  @Output() selectRunsOptionChosenEvent = new EventEmitter<SelectRunsOption>();
```

SelectRunsOptions enum class definition:
```
export enum SelectRunsOption {
  All = 'ALL',
  None = 'NONE',
  Completed = 'COMPLETED',
  Running = 'RUNNING',
  Scheduled = 'SCHEDULED'
}
```

hirokiterashima commented on src/app/teacher/select-runs-controls/select-runs-controls.component.ts:
```diff
@@ -0,0 +1,126 @@
+    this.selectedSomeRuns = this.numSelectedRuns !== 0 && !this.selectedAllRuns;
+  }
+
+  protected selectAllRunsCheckboxClicked(): void {
+    this.selectRunsOptionChosenEvent.emit(
+      this.selectedAllRuns || this.selectedSomeRuns ? 'none' : 'all'
+    );
+  }
+
+  protected selectRunsOptionChosen(value: string): void {
+    this.selectRunsOptionChosenEvent.emit(value);
+  }
```
Emit SelectRunsOption enum value?
```suggestion
  protected selectAllRunsCheckboxClicked(): void {
    this.selectRunsOptionChosenEvent.emit(
      this.selectedAllRuns || this.selectedSomeRuns ? SelectRunsOption.None : SelectRunsOption.All
    );
  }

  protected selectRunsOptionChosen(value: string): void {
    this.selectRunsOptionChosenEvent.emit(value as SelectRunsOption);
  }
```

hirokiterashima commented on src/app/teacher/select-runs-controls/select-runs-controls.component.html:
```diff
@@ -0,0 +1,53 @@
+      <button mat-menu-item (click)="selectRunsOptionChosen('none')" i18n>None</button>
+      <button mat-menu-item (click)="selectRunsOptionChosen('completed')" i18n>Completed</button>
+      <button mat-menu-item (click)="selectRunsOptionChosen('running')" i18n>Running</button>
+      <button mat-menu-item (click)="selectRunsOptionChosen('scheduled')" i18n>Scheduled</button>
```
Use enum values?
```suggestion
      <button mat-menu-item (click)="selectRunsOptionChosen('ALL')" i18n>All</button>
      <button mat-menu-item (click)="selectRunsOptionChosen('NONE')" i18n>None</button>
      <button mat-menu-item (click)="selectRunsOptionChosen('COMPLETED')" i18n>Completed</button>
      <button mat-menu-item (click)="selectRunsOptionChosen('RUNNING')" i18n>Running</button>
      <button mat-menu-item (click)="selectRunsOptionChosen('SCHEDULED')" i18n>Scheduled</button>
```

hirokiterashima reviewed (changes requested):
I added some code suggestions to prevent adding new code duplication issues in the TeacherRunListComponent spec file. I referenced [Jasmine style guide](https://github.com/CareMessagePlatform/jasmine-styleguide).

I also think using a new SelectRunsOption enum to represent available options will make the code clearer, so I added some suggestions inline for that as well. Lmk what you think or have any questions.

geoffreykwan committed:
feat(Archive Run): Use select runs option enum

#1012

geoffreykwan committed:
test(Archive Run): Clean up test and add missing new line

#1012

geoffreykwan committed:
test(Archive Run): Clean up test

#1012

geoffreykwan committed:
Merge branch 'develop' into issue-1012-allow-teacher-to-archive-runs

hirokiterashima reviewed (approved):
LGTM

breity reviewed (approved):
Looks good!

geoffreykwan committed:
fix(Archive Run): Display first 10 runs immediately

#1012

geoffreykwan committed:
feat(Archive Run): Only show number of runs after all runs are loaded

#1012

hirokiterashima commented:
:tada: This PR is included in version 5.112.0 :tada:

The release is available on [GitHub release](https://github.com/WISE-Community/WISE-Client/releases/tag/v5.112.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
